packages/contracts/deploy/30_upgrade_repo/31b_upgrade_and_reinitialize_repo.ts

```diff
@@ -0,0 +1,56 @@
+import {upgradeCommon as commonUpgradeSteps, skipUpgrade} from './_common';
+import {PLUGIN_REPO_PERMISSIONS} from '@aragon/osx-commons-sdk';
+import {BytesLike} from 'ethers';
+import {DeployFunction} from 'hardhat-deploy/types';
+import {HardhatRuntimeEnvironment} from 'hardhat/types';
+
+/**
+ * Upgrades the plugin repo to the latest implementation and reinitializes the proxy.
+ * @param {HardhatRuntimeEnvironment} hre
+ */
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const {deployer, pluginRepo, latestPluginRepoImplementation, current} =
+    await commonUpgradeSteps(hre);
+
+  // Define the `_initData` arguments
+  const initData: BytesLike[] = [];
+
+  // Encode the call to `function initializeFrom(uint8[3] calldata _previousProtocolVersion, bytes calldata _initData)` with `initData`.
+  const initializeFromCalldata =
+    latestPluginRepoImplementation.interface.encodeFunctionData(
+      // Re-initialization will happen through a call to `function initializeFrom(uint8[3] calldata _previousProtocolVersion, bytes calldata _initData)`
+      // that Aragon will add to the `PluginRepo` contract once it's required.
+      'initializeFrom',
+      [current, initData]
+    );
+
+  // Check if deployer has the permission to upgrade the plugin repo
+  if (
+    await pluginRepo.isGranted(
+      pluginRepo.address,
+      deployer.address,
+      PLUGIN_REPO_PERMISSIONS.UPGRADE_REPO_PERMISSION_ID,
+      []
+    )
+  ) {
+    // Use `upgradeToAndCall` to reinitialize the new `PluginRepo` implementation after the update.
+    if (initializeFromCalldata.length > 0) {
+      await pluginRepo.upgradeToAndCall(
+        latestPluginRepoImplementation.address,
+        initializeFromCalldata
+      );
+    } else {
+      throw Error(
+        `Initialization data is missing for 'upgradeToAndCall'. Stopping repo upgrade and reinitialization...`
+      );
+    }
+  } else {
+    throw Error(
+      `The new version cannot be published because the deployer ('${deployer.address}')
+      is lacking the ${PLUGIN_REPO_PERMISSIONS.UPGRADE_REPO_PERMISSION_ID} permission.`
+    );
+  }
+};
+export default func;
+func.tags = ['UpgradeAndReinitializeRepo'];
```

Comment:
Both of these files will be called by default (as you must specify the correct tags to use just one) meaning the default behaviour will be to upgrade without reinitialization. This is good, but make sure it's documented somewhere.

Reply:
Oh! Good catch! I think for it to be safe, it should be the other way around.
Otherwise, a user could be left with a contract that is not re-initialized, meaning that everybody could go and reinitialize it